protocols/relay/src/priv_client/handler.rs

```diff
@@ -458,38 +351,167 @@ impl ConnectionHandler for Handler {
             return Poll::Ready(ConnectionHandlerEvent::Close(err));
         }
 
+        // Circuit connections
+        if let Poll::Ready(worker_res) = self.circuit_connection_futs.poll_unpin(cx) {
+            let res = match worker_res {
+                Ok(r) => r,
+                Err(Timeout { .. }) => {
+                    return Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Timeout));
+                }
+            };
+
+            let opt = match res {
+                Ok(Some(outbound_hop::Output::Circuit { limit })) => {
+                    Some(ConnectionHandlerEvent::NotifyBehaviour(
+                        Event::OutboundCircuitEstablished { limit },
+                    ))
+                }
+                Ok(None) => None,
+                Err(err) => {
+                    let res = match err {
```

Comment:
Can collapse these `match`es.

```suggestion
                Err(outbound_hop::UpgradeError::CircuitFailed(e)) => {
```